Pegasus/include/geometry.hpp

```diff
@@ -164,1640 +184,668 @@ namespace geometry {
+        std::array<Vector3, 6> aBoxFaces, bBoxFaces;
+        std::vector<Vector3> separatingAxes;
+        real penetration = 0;
+    };
 
-    public:
-        IntersectionQueries(Plane const* a, Triangle const* b)
-            : Base(a, b)
-        {
-            if (initialized) {
-                calculate();
-            }
-        }
+    template < typename ShapeA, typename ShapeB >
+    void initialize(SimpleShape const * a, SimpleShape const * b, CacheBase  * cache);
 
-        bool overlap() const
-        {
-            if (initialized) {
-                auto const d = b->getCenterOfMass() * aNormal;
+    template < typename ShapeA, typename ShapeB >
+    bool overlap(SimpleShape const * a, SimpleShape const * b, CacheBase  * cache);
 
-                return (projections[0] < d) || (projections[1] < d) || (projections[2] < d);
-            }
+    template < typename ShapeA, typename ShapeB >
+    Vector3 calculateContactNormal(SimpleShape const * a, SimpleShape const * b, CacheBase  * cache);
 
-            return false;
-        }
+    template < typename ShapeA, typename ShapeB >
+    real calculatePenetration(SimpleShape const * a, SimpleShape const * b, CacheBase  * cache);
 
-        Vector3 calculateContactNormal() const
-        {
-            if (initialized) {
-                return b->getNormal().unit();
-            }
+    //Plane, Plane
+    template <>
+    inline void initialize<Plane, Plane>(const SimpleShape *a, const SimpleShape *b, CacheBase *cache)
+    {
+        auto const * p1 = static_cast<Plane const *>(a);
+        auto const * p2 = static_cast<Plane const *>(b);
+        auto * c = static_cast<Cache<Plane, Plane>*>(cache);
 
-            return {};
-        }
+        c->aNormal = p1->getNormal();
+        c->bNormal = p2->getNormal();
+        c->crossProduct = c->aNormal % c->bNormal;
+    }
 
-        real calculatePenetration() const
-        {
-            if (initialized) {
-                return *std::min_element(projections.begin(), projections.end());
-            }
+    template <>
+    inline bool overlap<Plane, Plane>(const SimpleShape *a, const SimpleShape *b, CacheBase *cache)
+    {
+        auto * c = static_cast<Cache<Plane, Plane>*>(cache);
+        return c->crossProduct.squareMagnitude() != static_cast<real>(0);
+    }
 
-            return 0;
-        }
+    template <>
+    inline Vector3 calculateContactNormal<Plane, Plane>(const SimpleShape *a, const SimpleShape *b, CacheBase *cache)
+    {
+        auto * c = static_cast<Cache<Plane, Plane>*>(cache);
+        return c->bNormal;
+    }
 
-    private:
-        void calculate()
-        {
-            b->getAxes(i, j, k);
-            aNormal = a->getNormal();
-            projections = { i * aNormal, j * aNormal, k * aNormal };
-        }
-    };
+    template <>
+    inline real calculatePenetration<Plane, Plane>(const SimpleShape *a, const SimpleShape *b, CacheBase *cache)
```

Comment:
Plane-Plane penetration depth is an invalid query, plane-plane intersection has an infinitely large depth. So it returns maximum value.